### PR TITLE
correction in backward function of log softmax layer

### PR DESCRIPTION
**Pull request by mrityunjay-tripathi**
Hi everyone,
The backward function in the LogSoftmax layer seems to have some error as neither it satisfies mathematical equation nor it passes Jacobian Test (i.e. producing large error). The deduction for derivative of LogSoftmax function is:
<p align="center">

</p>
...but the current implementation is:

https://github.com/mlpack/mlpack/blob/fded9ee333bfbac179e9826bdfe4a3b093ef67f2/src/mlpack/methods/ann/layer/log_softmax_impl.hpp#L64-L72

which doesn't look like the correct equation.
The error in the Jacobian Test for the new backward function is around 9.5e-05, but the error in Jacobian Test for the current backward function is around 0.98.

### CODE REVIEW

File: src/mlpack/methods/ann/layer/log_softmax_impl.hpp [hunk 1]
```diff
@@ -19,7 +19,7 @@ namespace mlpack {
 namespace ann /** Artificial Neural Network. */ {
 
 template<typename InputDataType, typename OutputDataType>
-LogSoftMax<InputDataType, OutputDataType>::LogSoftMax()
+LogSoftMax<InputDataType, OutputDataType>::LogSoftMax() : deterministic(0)
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/log_softmax_impl.hpp:22]: I don't think you need the use of `deterministic`. The `input` to `Backward` layer is the output obtained at the end of the `Forward` pass.

File: src/mlpack/methods/ann/layer/log_softmax_impl.hpp [hunk 2]
```diff
@@ -59,16 +59,21 @@ void LogSoftMax<InputDataType, OutputDataType>::Forward(
...
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
 void LogSoftMax<InputDataType, OutputDataType>::Backward(
-    const arma::Mat<eT>& input,
+    const arma::Mat<eT>& /* input */,
     const arma::Mat<eT>& gy,
     arma::Mat<eT>& g)
 {
-  g = arma::exp(input) + gy;
+  g = gy - arma::exp(y);
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/log_softmax_impl.hpp:76]: @zoq If I'm not mistaken, we do pass gradients from higher to lower layers with the negative multiplied?

File: src/mlpack/methods/ann/layer/log_softmax_impl.hpp (hunk 1 above)
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/log_softmax_impl.hpp:22]: I also thought that but when I saw the Jacobian Test function, 
https://github.com/mlpack/mlpack/blob/fded9ee333bfbac179e9826bdfe4a3b093ef67f2/src/mlpack/tests/ann_test_tools.hpp#L84-L96
...then I thought of doing this.

**Comment by mrityunjay-tripathi** [comment]
A lot of other tests are failing due to this. It's scary :(

File: src/mlpack/methods/ann/layer/log_softmax_impl.hpp (hunk 1 above)
Comment (ShikharJ) [reply src/mlpack/methods/ann/layer/log_softmax_impl.hpp:22]: I'm not sure what the `JacobianTest` is doing, maybe @zoq can help with this.

Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/log_softmax_impl.hpp:22]: Jacobian Test computes the derivative by using the first principle,
 ```f'(x) = lim (f(x + h) - f(x - h)) / 2h```and then comparing it with backward function. It is really confusing how other tests passed based on this. 😤

Comment (zoq) [reply src/mlpack/methods/ann/layer/log_softmax_impl.hpp:22]: Probably because the threshold is to high.

**Comment by zoq** [comment]
My first guess is that we have to multiply with -1 if used with `NegativeLogLikelihood`.

File: src/mlpack/methods/ann/layer/log_softmax_impl.hpp (hunk 1 above)
Comment (ShikharJ) [reply src/mlpack/methods/ann/layer/log_softmax_impl.hpp:22]: The threshold for functions where this is tested is around `1e-3` I think. Not sure it qualifies as high.

**Comment by mrityunjay-tripathi** [comment]
>My first guess is that we have to multiply with -1 if used with NegativeLogLikelihood.

Yeah. Also, I think that is shown in the unit-testing.
https://github.com/mlpack/mlpack/blob/fded9ee333bfbac179e9826bdfe4a3b093ef67f2/src/mlpack/tests/ann_layer_test.cpp#L1497-L1502
Probably it doesn't need to be multiplied intrinsically in LogSoftmax?

**Comment by mrityunjay-tripathi** [comment]
Some answers that I got:
1. The [mathematical NLL Loss](https://arxiv.org/abs/1804.10690) is split into two layers in the actual implementation (one LogSoftmax layer and other NLL Loss layer).
2. LogSoftmax layer computes the log of softmax of the input. NLL Loss accepts the output of the LogSoftmax layer (log-probabilities), sums it up and puts negative sign (`negative_log_likelihood_impl.hpp` line number 42 and 62).

Questions I am stuck with:
1. Do we still need to introduce minus sign in LogSoftmax `Backward` function? Why?
2. The `Backward` function of LogSoftmax is not in line with the [equation](https://math.stackexchange.com/a/2340848/506476) then how gradient tests like [these](https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/ann_layer_test.cpp#L84) are successful?
3. Why there is `input` (not `activation`) used in `JacobianTest`? And If `JacobianTest` is fine then why I am getting much less error (~9.5e-05) compared to error in current implementation (~0.98)?
4. Should we ignore `JacobianTest` completely here? 🤔

File: src/mlpack/methods/ann/layer/log_softmax_impl.hpp (hunk 2 above)
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/log_softmax_impl.hpp:76]: @ShikharJ @zoq: Is there a convention we should follow? The `HardTanH` layer test fails on Jacobian Test when activations are fed to the `Backward` function. And its the only layer that failed JacobianTest when I fed the `output` to `Backward` function in place of `input`. Should we add an argument to the JacobianTest whether we want to feed `input` or `activation` to the `Backward` function?

**Comment by mrityunjay-tripathi** [comment]
Ok. I think https://discuss.pytorch.org/t/does-nllloss-handle-log-softmax-and-softmax-in-the-same-way/8835 explains it. I'm really sorry @zoq and @ShikharJ for this trouble. I think I should close this PR.
